[android][ci] Fix CI packaging headers to aar

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40442 [android][ci] Fix CI packaging headers to aar**

Problem:
Nightly builds do not include libtorch headers as local build.
The reason is that on docker images path is different than local path when building with `scripts/build_pytorch_android.sh`

Solution:
Introducing gradle property to be able to specify it and add its specification to gradle build job and snapshots publishing job which run on the same docker image.

Test:
ci-all jobs check https://github.com/pytorch/pytorch/pull/40443
checking that gradle build will result with headers inside aar

Successful build of aar with headers: https://app.circleci.com/pipelines/github/pytorch/pytorch/184669/workflows/c5c5409f-26c8-43dc-8c33-126550f04aab/jobs/5983465/artifacts

Differential Revision: [D22190955](https://our.internmc.facebook.com/intern/diff/D22190955)

IvanKobzarev committed:
[android][ci] Fix CI packaging headers to aar

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android][ci] Fix CI packaging headers to aar"


Problem:
Nightly builds do not include libtorch headers as local build.
The reason is that on docker images path is different than local path when building with `scripts/build_pytorch_android.sh`

Solution:
Introducing gradle property to be able to specify it and add its specification to gradle build job and snapshots publishing job which run on the same docker image.

Test:
ci-all jobs check https://github.com/pytorch/pytorch/pull/40443
checking that gradle build will result with headers inside aar

Differential Revision: [D22190955](https://our.internmc.facebook.com/intern/diff/D22190955)

[ghstack-poisoned]

ljk53 commented on android/pytorch_android/build.gradle:
```diff
@@ -85,6 +85,12 @@ task sourcesJar(type: Jar) {
+  }
+
   def abi = ABI_FILTERS.split(",")[0]
   return "$rootDir/../build_android_$abi/install/include";
```
When do we hit this default value? Is it for `scripts/build_pytorch_android.sh`?

Can we always pass in this directory explicitly? When I refactored the script recently, I added the ability to override the `BUILD_ROOT` for build_pytorch_android.sh, i.e. `build_android_$abi` can now be in places like pytorch/build_pytorch_android_custom/build_android_x86/... - see my PR #40199 

In general I feel it would be more robust to reduce the hardcoded path constants in codebase :)

IvanKobzarev replied:
Yes, default value is for local build with `scripts/build_pytorch_android.sh`, that will also needed for `bintrayUpload` task for release publishing.

Yes, I agree, with hardcoding of paths we have to integrate gradle changes twice:

As we have different paths locally and on CI, on CI we have variations like `~/workspace/build_android_install_$abi/install/include` (_install_ in the middle) which is copied from abi docker image.

For selective build we do not have exported functions in binarys and we do not need headers in the package, am I right that custom build is the same as selective build that we also do not need to add headers inside?

IvanKobzarev replied:
I will try to remove hardcoded path for local build using the same -PLIBTORCH_HOME

IvanKobzarev committed:
Update on "[android][ci] Fix CI packaging headers to aar"


Problem:
Nightly builds do not include libtorch headers as local build.
The reason is that on docker images path is different than local path when building with `scripts/build_pytorch_android.sh`

Solution:
Introducing gradle property to be able to specify it and add its specification to gradle build job and snapshots publishing job which run on the same docker image.

Test:
ci-all jobs check https://github.com/pytorch/pytorch/pull/40443
checking that gradle build will result with headers inside aar

Differential Revision: [D22190955](https://our.internmc.facebook.com/intern/diff/D22190955)

[ghstack-poisoned]

ljk53 commented on android/pytorch_android/build.gradle:
```diff
@@ -85,8 +85,7 @@ task sourcesJar(type: Jar) {
 def getLibtorchHeadersDir() {
-  def abi = ABI_FILTERS.split(",")[0]
-  return "$rootDir/../build_android_$abi/install/include";
+  return property('LIBTORCH_HEADERS')
```
Took a closer look, maybe if we change the default value to 
`pytorch_android/src/main/cpp/libtorch_include/$abi` it will work for all cases? as it's a reasonable assumption to make that symbolic link to build_android/install/include must have been set at earlier build steps...

Sorry I didn't see this earlier - it's an optional change :)

IvanKobzarev committed:
Update on "[android][ci] Fix CI packaging headers to aar"


Problem:
Nightly builds do not include libtorch headers as local build.
The reason is that on docker images path is different than local path when building with `scripts/build_pytorch_android.sh`

Solution:
Introducing gradle property to be able to specify it and add its specification to gradle build job and snapshots publishing job which run on the same docker image.

Test:
ci-all jobs check https://github.com/pytorch/pytorch/pull/40443
checking that gradle build will result with headers inside aar

Successful build of aar with headers: https://app.circleci.com/pipelines/github/pytorch/pytorch/184669/workflows/c5c5409f-26c8-43dc-8c33-126550f04aab/jobs/5983465/artifacts

Differential Revision: [D22190955](https://our.internmc.facebook.com/intern/diff/D22190955)

[ghstack-poisoned]
